core/trino-main/src/main/java/io/trino/sql/ir/NullIf.java

```diff
@@ -32,6 +32,8 @@ public record NullIf(Expression first, Expression second)
     {
         requireNonNull(first, "first is null");
         requireNonNull(second, "second is null");
+
+        // TODO: verify that first and second can be coerced to the same type
```

Comment:
you could add the `Optional<ResolvedFunction>` for the coercion and then the verification is easier.. also you likely need the resolved equals function here